Add network SLO (#19)
appuio/component-openshift4-slos: docs/modules/ROOT/pages/runbooks/network.adoc

= Network SLOs

include::partial$runbooks/contribution_note.adoc[]

[[canary]]
== Cluster Network Packet Loss

=== icon:glasses[] Overview

> The SLI measures the percentage of dropped ICMP ping messages between canary pods on ever worker node.

Reviewer: ```suggestion
The SLI measures the percentage of dropped ICMP ping messages between canary pods on every worker node.
```

> Every second on every worker node a canary pod pings canary pods on ever other worker node and reports the success as a Prometheus metric.

Reviewer: ```suggestion
Every second on each worker node a canary pod pings canary pods on every other worker node and reports the success as a Prometheus metric.
```

> If this SLI results in an alert, it means the some network packets between pods are lost.

Reviewer: ```suggestion
If this SLI results in an alert, it means that some network packets between pods are lost.
```

include::partial$runbooks/alert_types.adoc[]

=== icon:bug[] Steps for debugging

Packet loss can have multiple root causes and the impact of this alert might vary.
Given that the cluster was able to send this alert, a complete outage of the network is unlikely.

In any case, first try to connect to the affected cluster and to open the Prometheus UI.
If that works successfully you can run the following query to see between which pods and nodes network packets are lost.

[source, promql]
----
sum(rate(ping_rtt_seconds_count{reason="lost"}[5m])) by (target, source)
* on (target) group_left(target_node)
   label_replace(
    label_replace(
     max(kube_pod_info{pod_ip!~""}) by (pod_ip, node),
    "target", "$1", "pod_ip", "(.+)"),
   "target_node", "$1", "node", "(.+)")
* on (source) group_left(source_node)
   label_replace(
    label_replace(
     max(kube_pod_info{pod_ip!~""}) by (pod_ip, node),
    "source", "$1", "pod_ip", "(.+)"),
   "source_node", "$1", "node", "(.+)")
> 0
----

> This query should one or more time series, indicating packet loss between canary pods an which nodes they run on.

Reviewer: ```suggestion
This query should return one or more time series, indicating packet loss between canary pods and which nodes they run on.
```

From that you should be able to see which nodes are effected.
If the target or source node is always the same, this is most likely an issue with that specific node, maybe it just crashed and this isn't a network issue after all.

==== Replicate Issue

If the issue isn't obvious, it's a good idea to first replicate the issue.

First look at the logs of one of the source canary pods reporting packet loss.
Find the relevant pod from the reported source IP.

[source, shell]
----
kubectl -n appuio-network-canary get pod -o wide
kubectl -n appuio-network-canary logs ${CANARY_POD}
----

> In the logs you should be able to see if the issue is actually DNS related or if there is any other unexpected issues.

Reviewer: ```suggestion
In the logs you should be able to see if the issue is actually DNS related or if there are any other unexpected issues.
```

> If DNS isn't working properly, service discovery of the network canary won't work, which can lead to miss reported packet loss.

Reviewer: ```suggestion
If DNS isn't working properly, service discovery of the network canary won't work, which can lead to misreported packet loss.
```
